plugins/TapTempo/TapTempo.cpp

```diff
@@ -0,0 +1,142 @@
+/*
+ * TapTempo.cpp - plugin to count beats per minute
+ *
+ *
+ * Copyright (c) 2022 sakertooth <sakertooth@gmail.com>
+ * 
+ * 
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "TapTempo.h"
+
+#include <cmath>
+
+#include <QFont>
+#include <QLabel>
+#include <QVBoxLayout>
+
+#include "embed.h"
+#include "plugin_export.h"
+
+extern "C"
+{
+Plugin::Descriptor PLUGIN_EXPORT taptempo_plugin_descriptor =
+{
+		STRINGIFY( PLUGIN_NAME ),
+		"Tap Tempo",
+		QT_TRANSLATE_NOOP( "PluginBrowser",
+							"Tap to the beat" ),
+		"sakertooth <sakertooth@gmail.com>",
+		0x0100,
+		Plugin::Tool,
+		new PluginPixmapLoader("logo"),
+		nullptr,
+		nullptr
+} ;
+
+// neccessary for getting instance out of shared lib
+PLUGIN_EXPORT Plugin * lmms_plugin_main(Model *, void *)
+{
+	return new TapTempo;
+}
+}
+
+TapTempo::TapTempo() : 
+	ToolPlugin(&taptempo_plugin_descriptor, nullptr)
+{
+}
+
+QString TapTempo::nodeName() const
+{
+	return taptempo_plugin_descriptor.name;
+}
+
+TapTempoView::TapTempoView(ToolPlugin * _tool) :
+	ToolPluginView(_tool), m_bpmAverage(0), m_numTaps(0), m_keyDown(false)
+{
+	setFixedSize(200, 200);
+	m_bpmButton = new QPushButton;
+	m_bpmButton->setText("0");
+	m_bpmButton->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+
+	QFont font = m_bpmButton->font();
+	font.setPointSize(25);
+	m_bpmButton->setFont(font);
+	
+	QVBoxLayout * layout = new QVBoxLayout(this);
+	layout->setAlignment(Qt::AlignCenter);
+	layout->addWidget(m_bpmButton, Qt::AlignCenter);
+	
+	connect(m_bpmButton, &QPushButton::pressed, this, &TapTempoView::onBpmClick);
+
+	hide();
+	if(parentWidget())
+	{
+		parentWidget()->hide();
+		parentWidget()->layout()->setSizeConstraint(QLayout::SetFixedSize);
+
+		Qt::WindowFlags flags = parentWidget()->windowFlags();
+		flags |= Qt::MSWindowsFixedSizeDialogHint;
+		flags &= ~Qt::WindowMaximizeButtonHint;
+		parentWidget()->setWindowFlags(flags);
+	}
+}
+
+void TapTempoView::onBpmClick()
```

Comment:
Why not
- Grab `thisTime` (`currentTime`) immediately at the top of the function, then use that later
- Store `firstTime` and `latestTime`, use (latest - first) / (taps - 1) instead of m_bpmAverage

Something like this should work, right? (latestTime should be initialized to a long time ago)
```
thisTime = now
if (thisTime - latestTime > 2 seconds)
    // 2 seconds passed, reset calculation
    taps = 1
    firstTime = thisTime
    latestTime = thisTime
else
    // register latest tap
    taps++
    latestTime = thisTime
    
average = 60 * (taps - 1) / max(DBL_MIN, latestTime - thisTime)
setText( average )
```

Reply:
My interpretation may be wrong, but I am not too sure I fully agree with this implementation.

The reason why the current implementation uses an average BPM and wave period rather than just the most current one is because very small changes in wave periods make the BPM less stable and shift more frequently. So, it is, or at least how I see it, more favorable to consider the BPM over each tap as time passes.

Say, for example, the difference in time between 2 taps was 0.2587 seconds and another was 0.2892 seconds. Knowing that the number of beats per second (Hz) is equal to 1 divided by the wave period in seconds:
- Beats per minute given that the distance between each beat is 0.2587 seconds: 60/0.2587 = ~231.929
- Beats per minute given that the distance between each beat is 0.2892 seconds: 60/0.2892 = ~207.469

The difference in seconds between each beat was 0.0305 (30.5 ms, basically no difference in time to the average person) but the BPM drastically changed by 24.46. If instead we were to take an average BPM over time, we would get a better understanding as to what the BPM is.

Seeing `average = 60 * (taps - 1) / max(DBL_MIN, latestTime - thisTime)`, what would happen if `taps` was > 2?

Assuming `latestTime - thisTime` was, say, 0.4893 and the number of taps was 5, wouldn't this expression get the number of beats for every 4 minutes instead of 1 minute?

I may be wrong however.

Reply:
Sorry, I misstyped. The average should use `firstTime - thisTime` as the period. This ensures that floating point inaccuracies don't accumulate & the tempo gets more accurate with each click. 

> Say, for example, the difference in time between 2 taps was 0.2587 seconds and another was 0.2892 seconds.

With the fixed equation, the first 2 taps would give 231.93 as you say, and then the third would update to 219.02

`60 * (0.2587 + 0.2892)/2`

Reply:
I understand now. This is counting the number of taps that happen in the increasing time duration, rather than taking an average of each period. The code is more compact and it looks like the tempo becomes accurate faster. Thanks.